File: client/src/websocket_client.cpp
#include "websocket_client.hpp"
#include <iostream>
#include <random>
#include <sstream>
#include <iomanip>
#include <fstream>

namespace deeper_hub {

WebSocketClient::WebSocketClient(const std::string& url, const std::string& auth_token, bool use_tls)
    : url_(url)
    , auth_token_(auth_token)
    , use_tls_(use_tls)
    , connected_(false)
    , authenticated_(false)
    , reconnecting_(false)
    , response_received_(false)
    , heartbeat_running_(false)
{
    // Inicializa o cliente WebSocket apropriado (TLS ou não)
    if (use_tls_) {
        wss_client_ = std::make_unique<websocket_client_tls>();
        
        // Configuração do cliente TLS
        wss_client_->set_access_channels(websocketpp::log::alevel::all);
        wss_client_->clear_access_channels(websocketpp::log::alevel::frame_payload);
        wss_client_->set_error_channels(websocketpp::log::elevel::all);
        
        // Inicializa ASIO
        wss_client_->init_asio();
        
        // Registra handlers
        wss_client_->set_open_handler([this](connection_hdl hdl) { this->on_open(hdl); });
        wss_client_->set_message_handler([this](connection_hdl hdl, message_ptr msg) { this->on_message(hdl, msg); });
        wss_client_->set_close_handler([this](connection_hdl hdl) { this->on_close(hdl); });
        wss_client_->set_fail_handler([this](connection_hdl hdl) { this->on_fail(hdl); });
        
        // Desativa verificação de certificado para desenvolvimento (remover em produção)
        wss_client_->set_tls_init_handler([](connection_hdl) {
            return websocketpp::lib::make_shared<boost::asio::ssl::context>(boost::asio::ssl::context::tlsv12);
        });
    } else {
        ws_client_ = std::make_unique<websocket_client>();
        
        // Configuração do cliente não-TLS
        ws_client_->set_access_channels(websocketpp::log::alevel::all);
        ws_client_->clear_access_channels(websocketpp::log::alevel::frame_payload);
        ws_client_->set_error_channels(websocketpp::log::elevel::all);
        
        // Inicializa ASIO
        ws_client_->init_asio();
        
        // Registra handlers
        ws_client_->set_open_handler([this](connection_hdl hdl) { this->on_open(hdl); });
        ws_client_->set_message_handler([this](connection_hdl hdl, message_ptr msg) { this->on_message(hdl, msg); });
        ws_client_->set_close_handler([this](connection_hdl hdl) { this->on_close(hdl); });
        ws_client_->set_fail_handler([this](connection_hdl hdl) { this->on_fail(hdl); });
    }
}

WebSocketClient::~WebSocketClient() {
    // Para o heartbeat
    stop_heartbeat();
    
    // Desconecta do servidor
    disconnect();
    
    // Limpa recursos
    if (use_tls_) {
        wss_client_->stop_perpetual();
    } else {
        ws_client_->stop_perpetual();
    }
}

bool WebSocketClient::connect() {
    if (connected_) {
        std::cout << "Cliente já está conectado" << std::endl;
        return true;
    }
    
    try {
        websocketpp::lib::error_code ec;
        
        if (use_tls_) {
            // Conecta usando TLS
            connection_ = wss_client_->get_connection(url_, ec);
            if (ec) {
                std::cerr << "Erro ao criar conexão: " << ec.message() << std::endl;
                return false;
            }
            
            wss_client_->connect(connection_);
            wss_client_->run();
        } else {
            // Conecta sem TLS
            connection_ = ws_client_->get_connection(url_, ec);
            if (ec) {
                std::cerr << "Erro ao criar conexão: " << ec.message() << std::endl;
                return false;
            }
            
            ws_client_->connect(connection_);
            ws_client_->run();
        }
        
        // Aguarda até que a conexão seja estabelecida
        auto start_time = std::chrono::steady_clock::now();
        while (!connected_ && 
               std::chrono::duration_cast<std::chrono::seconds>(
                   std::chrono::steady_clock::now() - start_time).count() < 10) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
        
        // Verifica se a conexão foi estabelecida
        if (!connected_) {
            std::cerr << "Timeout ao conectar ao servidor" << std::endl;
            return false;
        }
        
        // Aguarda até que a autenticação seja concluída
        start_time = std::chrono::steady_clock::now();
        while (!authenticated_ && 
               std::chrono::duration_cast<std::chrono::seconds>(
                   std::chrono::steady_clock::now() - start_time).count() < 10) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
        
        // Verifica se a autenticação foi concluída
        if (!authenticated_) {
            std::cerr << "Timeout na autenticação" << std::endl;
            disconnect();
            return false;
        }
        
        std::cout << "Conectado e autenticado com sucesso" << std::endl;
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Exceção ao conectar: " << e.what() << std::endl;
        return false;
    }
}

void WebSocketClient::disconnect() {
    if (!connected_) {
        return;
    }
    
    try {
        websocketpp::lib::error_code ec;
        
        if (use_tls_) {
            wss_client_->close(connection_, websocketpp::close::status::normal, "Desconexão normal", ec);
        } else {
            ws_client_->close(connection_, websocketpp::close::status::normal, "Desconexão normal", ec);
        }
        
        if (ec) {
            std::cerr << "Erro ao fechar conexão: " << ec.message() << std::endl;
        }
    } catch (const std::exception& e) {
        std::cerr << "Exceção ao desconectar: " << e.what() << std::endl;
    }
    
    connected_ = false;
    authenticated_ = false;
}

bool WebSocketClient::is_connected() const {
    return connected_;
}

bool WebSocketClient::is_authenticated() const {
    return authenticated_;
}

bool WebSocketClient::send_message(const json& message) {
    if (!connected_ || !authenticated_) {
        std::cerr << "Tentativa de enviar mensagem sem conexão ou autenticação" << std::endl;
        return false;
    }
    
    try {
        std::string message_str = message.dump();
        websocketpp::lib::error_code ec;
        
        if (use_tls_) {
            wss_client_->send(connection_, message_str, websocketpp::frame::opcode::text, ec);
        } else {
            ws_client_->send(connection_, message_str, websocketpp::frame::opcode::text, ec);
        }
        
        if (ec) {
            std::cerr << "Erro ao enviar mensagem: " << ec.message() << std::endl;
            return false;
        }
        
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Exceção ao enviar mensagem: " << e.what() << std::endl;
        return false;
    }
}

json WebSocketClient::wait_for_response(int timeout_ms) {
    std::unique_lock<std::mutex> lock(response_mutex_);
    
    // Reseta o estado de resposta
    response_received_ = false;
    last_response_ = json();
    
    // Aguarda pela resposta com timeout
    if (response_cv_.wait_for(lock, std::chrono::milliseconds(timeout_ms),
                             [this] { return response_received_; })) {
        return last_response_;
    } else {
        std::cerr << "Timeout ao aguardar resposta" << std::endl;
        return json();
    }
}

void WebSocketClient::set_message_callback(std::function<void(const json&)> callback) {
    message_callback_ = callback;
}

void WebSocketClient::set_connection_callback(std::function<void(bool)> callback) {
    connection_callback_ = callback;
}

void WebSocketClient::start_heartbeat(int interval_ms) {
    // Para o heartbeat atual se estiver rodando
    stop_heartbeat();
    
    // Inicia um novo thread de heartbeat
    heartbeat_running_ = true;
    heartbeat_thread_ = std::thread(&WebSocketClient::heartbeat_loop, this, interval_ms);
}

void WebSocketClient::stop_heartbeat() {
    // Para o thread de heartbeat se estiver rodando
    if (heartbeat_running_) {
        heartbeat_running_ = false;
        if (heartbeat_thread_.joinable()) {
            heartbeat_thread_.join();
        }
    }
}

void WebSocketClient::on_open(connection_hdl hdl) {
    std::cout << "Conexão estabelecida" << std::endl;
    connected_ = true;
    
    // Notifica o callback de conexão
    if (connection_callback_) {
        connection_callback_(true);
    }
    
    // Envia mensagem de join com autenticação
    join_channel();
}

void WebSocketClient::on_message(connection_hdl hdl, message_ptr msg) {
    try {
        // Converte a mensagem para JSON
        std::string payload = msg->get_payload();
        json data = json::parse(payload);
        
        std::cout << "Mensagem recebida: " << payload << std::endl;
        
        // Processa diferentes tipos de mensagens
        if (data.contains("event")) {
            std::string event_type = data["event"];
            
            if (event_type == "phx_reply" && data.value("ref", "") == "1") {
                // Resposta de autenticação
                if (data.contains("payload") && data["payload"].contains("status") && 
                    data["payload"]["status"] == "ok") {
                    std::cout << "Autenticação bem-sucedida" << std::endl;
                    authenticated_ = true;
                } else {
                    std::cerr << "Erro na autenticação: " << data.dump() << std::endl;
                }
            } else if (event_type == "heartbeat") {
                // Heartbeat recebido
                std::cout << "Heartbeat recebido" << std::endl;
            } else if (event_type == "phx_reply") {
                // Extrai a resposta do formato Phoenix
                json response_data = data.value("payload", json::object());
                
                // Se for uma resposta de operação de banco de dados
                if (response_data.is_object()) {
                    if (response_data.contains("response") && response_data["response"].is_object()) {
                        json response = response_data["response"];
                        if (response.contains("type") && response["type"] == "database_response") {
                            std::cout << "Resposta de operação de banco de dados recebida" << std::endl;
                            
                            // Notifica que uma resposta foi recebida
                            {
                                std::lock_guard<std::mutex> lock(response_mutex_);
                                last_response_ = response;
                                response_received_ = true;
                            }
                            response_cv_.notify_one();
                            
                            // Notifica o callback de mensagem
                            if (message_callback_) {
                                message_callback_(response);
                            }
                        } else if (response.contains("status")) {
                            std::cout << "Resposta recebida: " << response.dump() << std::endl;
                            
                            // Notifica que uma resposta foi recebida
                            {
                                std::lock_guard<std::mutex> lock(response_mutex_);
                                last_response_ = response;
                                response_received_ = true;
                            }
                            response_cv_.notify_one();
                            
                            // Notifica o callback de mensagem
                            if (message_callback_) {
                                message_callback_(response);
                            }
                        }
                    } else if (response_data.contains("status")) {
                        std::cout << "Resposta direta recebida: " << response_data.dump() << std::endl;
                        
                        // Notifica que uma resposta foi recebida
                        {
                            std::lock_guard<std::mutex> lock(response_mutex_);
                            last_response_ = response_data;
                            response_received_ = true;
                        }
                        response_cv_.notify_one();
                        
                        // Notifica o callback de mensagem
                        if (message_callback_) {
                            message_callback_(response_data);
                        }
                    }
                }
            }
        } else if (data.is_object() && data.contains("type") && data["type"] == "database_response") {
            // Formato antigo de resposta de banco de dados (compatibilidade)
            std::cout << "Resposta de operação de banco de dados recebida (formato antigo): " 
                     << data["operation"] << " - " << data["schema"] << std::endl;
            
            // Notifica que uma resposta foi recebida
            {
                std::lock_guard<std::mutex> lock(response_mutex_);
                last_response_ = data;
                response_received_ = true;
            }
            response_cv_.notify_one();
            
            // Notifica o callback de mensagem
            if (message_callback_) {
                message_callback_(data);
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Erro ao processar mensagem: " << e.what() << std::endl;
    }
}

void WebSocketClient::on_close(connection_hdl hdl) {
    std::cout << "Conexão fechada" << std::endl;
    connected_ = false;
    authenticated_ = false;
    
    // Notifica o callback de conexão
    if (connection_callback_) {
        connection_callback_(false);
    }
}

void WebSocketClient::on_fail(connection_hdl hdl) {
    std::cerr << "Falha na conexão" << std::endl;
    connected_ = false;
    authenticated_ = false;
    
    // Notifica o callback de conexão
    if (connection_callback_) {
        connection_callback_(false);
    }
}

void WebSocketClient::join_channel() {
    // Cria mensagem de join no formato Phoenix WebSocket
    json join_message = {
        {"topic", "websocket"},
        {"event", "phx_join"},
        {"payload", {{"auth_token", auth_token_}}},
        {"ref", "1"}
    };
    
    // Envia a mensagem
    std::string message_str = join_message.dump();
    websocketpp::lib::error_code ec;
    
    if (use_tls_) {
        wss_client_->send(connection_, message_str, websocketpp::frame::opcode::text, ec);
    } else {
        ws_client_->send(connection_, message_str, websocketpp::frame::opcode::text, ec);
    }
    
    if (ec) {
        std::cerr << "Erro ao enviar mensagem de join: " << ec.message() << std::endl;
    } else {
        std::cout << "Mensagem de join enviada" << std::endl;
    }
}

void WebSocketClient::heartbeat_loop(int interval_ms) {
    while (heartbeat_running_ && connected_) {
        // Aguarda pelo intervalo
        std::this_thread::sleep_for(std::chrono::milliseconds(interval_ms));
        
        if (!heartbeat_running_ || !connected_) {
            break;
        }
        
        // Cria mensagem de heartbeat
        json heartbeat_message = {
            {"topic", "websocket"},
            {"event", "heartbeat"},
            {"payload", json::object()},
            {"ref", generate_uuid()}
        };
        
        // Envia a mensagem
        send_message(heartbeat_message);
    }
}

std::string WebSocketClient::generate_uuid() {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> dis(0, 15);
    static std::uniform_int_distribution<> dis2(8, 11);
    
    std::stringstream ss;
    ss << std::hex;
    
    for (int i = 0; i < 8; i++) {
        ss << dis(gen);
    }
    ss << "-";
    
    for (int i = 0; i < 4; i++) {
        ss << dis(gen);
    }
    ss << "-4";
    
    for (int i = 0; i < 3; i++) {
        ss << dis(gen);
    }
    ss << "-";
    
    ss << dis2(gen);
    
    for (int i = 0; i < 3; i++) {
        ss << dis(gen);
    }
    ss << "-";
    
    for (int i = 0; i < 12; i++) {
        ss << dis(gen);
    }
    
    return ss.str();
}

} // namespace deeper_hub

File: test/websocket_client/websocket_client.py
#!/usr/bin/env python3
"""
Cliente WebSocket para testar a comunicação com o servidor Deeper_Hub.

Este script:
- Conecta-se ao servidor WebSocket
- Envia mensagens de autenticação
- Envia heartbeats periódicos
- Processa mensagens recebidas
"""

import json
import time
import threading
import websocket
import logging

# Configuração de logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s'
)
logger = logging.getLogger('websocket_client')

class DeeperHubClient:
    """Cliente WebSocket para o Deeper_Hub."""
    
    def __init__(self, url="ws://localhost:4000/socket/websocket", auth_token="test_token"):
        """
        Inicializa o cliente WebSocket.
        
        Args:
            url: URL do servidor WebSocket
            auth_token: Token de autenticação
        """
        self.url = url
        self.auth_token = auth_token
        self.ws = None
        self.connected = False
        self.should_reconnect = True
        self.heartbeat_thread = None
        
    def connect(self):
        """Conecta ao servidor WebSocket."""
        try:
            # Configuração do WebSocket
            websocket.enableTrace(True)
            self.ws = websocket.WebSocketApp(
                self.url,
                on_open=self.on_open,
                on_message=self.on_message,
                on_error=self.on_error,
                on_close=self.on_close
            )
            
            # Inicia a conexão em uma thread separada
            self.ws_thread = threading.Thread(target=self.ws.run_forever)
            self.ws_thread.daemon = True
            self.ws_thread.start()
            
            logger.info(f"Conectando ao servidor: {self.url}")
            return True
        except Exception as e:
            logger.error(f"Erro ao conectar: {e}")
            return False
            
    def on_open(self, ws):
        """Callback quando a conexão é estabelecida."""
        logger.info("Conexão estabelecida")
        self.connected = True
        
        # Envia mensagem de join com autenticação
        self.join_channel()
        
        # Inicia o heartbeat
        self.start_heartbeat()
        
    def on_message(self, ws, message):
        """Callback quando uma mensagem é recebida."""
        try:
            logger.info(f"Mensagem recebida: {message}")
            data = json.loads(message)
            
            # Processa diferentes tipos de mensagens
            if 'event' in data:
                event_type = data['event']
                if event_type == 'phx_reply' and data.get('ref') == '1':
                    if data.get('payload', {}).get('status') == 'ok':
                        logger.info("Autenticação bem-sucedida")
                    else:
                        logger.error(f"Erro na autenticação: {data}")
                elif event_type == 'heartbeat_response':
                    logger.info("Heartbeat recebido")
                else:
                    logger.info(f"Evento recebido: {event_type}")
        except Exception as e:
            logger.error(f"Erro ao processar mensagem: {e}")
            
    def on_error(self, ws, error):
        """Callback quando ocorre um erro."""
        logger.error(f"Erro na conexão: {error}")
        
    def on_close(self, ws, close_status_code, close_msg):
        """Callback quando a conexão é fechada."""
        logger.info(f"Conexão fechada: {close_status_code} - {close_msg}")
        self.connected = False
        
        # Tenta reconectar se necessário
        if self.should_reconnect:
            logger.info("Tentando reconectar em 5 segundos...")
            time.sleep(5)
            self.connect()
            
    def join_channel(self):
        """Envia mensagem para entrar no canal."""
        join_message = {
            "topic": "websocket",
            "event": "phx_join",
            "payload": {"auth_token": self.auth_token},
            "ref": "1"
        }
        self.send_message(join_message)
        logger.info("Mensagem de join enviada")
        
    def send_heartbeat(self):
        """Envia mensagem de heartbeat."""
        heartbeat_message = {
            "topic": "websocket",
            "event": "heartbeat",
            "payload": {"timestamp": int(time.time())},
            "ref": str(int(time.time()))
        }
        self.send_message(heartbeat_message)
        logger.debug("Heartbeat enviado")
        
    def start_heartbeat(self):
        """Inicia o envio periódico de heartbeats."""
        def heartbeat_loop():
            while self.connected:
                self.send_heartbeat()
                time.sleep(15)  # Envia heartbeat a cada 15 segundos
                
        self.heartbeat_thread = threading.Thread(target=heartbeat_loop)
        self.heartbeat_thread.daemon = True
        self.heartbeat_thread.start()
        
    def send_message(self, message):
        """Envia uma mensagem para o servidor."""
        if self.connected:
            try:
                self.ws.send(json.dumps(message))
                return True
            except Exception as e:
                logger.error(f"Erro ao enviar mensagem: {e}")
                return False
        else:
            logger.warning("Tentativa de enviar mensagem sem conexão")
            return False
            
    def send_custom_message(self, event, payload):
        """Envia uma mensagem personalizada para o servidor."""
        message = {
            "topic": "websocket",
            "event": event,
            "payload": payload,
            "ref": str(int(time.time()))
        }
        return self.send_message(message)
        
    def disconnect(self):
        """Desconecta do servidor."""
        self.should_reconnect = False
        if self.ws:
            self.ws.close()
            logger.info("Desconectado do servidor")

def main():
    """Função principal."""
    client = DeeperHubClient()
    
    # Conecta ao servidor
    if client.connect():
        try:
            # Mantém o programa em execução
            while True:
                cmd = input("\nComandos disponíveis:\n"
                           "1. Enviar mensagem personalizada\n"
                           "2. Enviar heartbeat\n"
                           "3. Sair\n"
                           "Digite o número do comando: ")
                
                if cmd == "1":
                    event = input("Digite o tipo de evento: ")
                    payload_str = input("Digite o payload (JSON): ")
                    try:
                        payload = json.loads(payload_str)
                        client.send_custom_message(event, payload)
                    except json.JSONDecodeError:
                        logger.error("Payload inválido. Use formato JSON.")
                
                elif cmd == "2":
                    client.send_heartbeat()
                    logger.info("Heartbeat enviado manualmente")
                
                elif cmd == "3":
                    break
                
                else:
                    logger.warning("Comando inválido")
                    
        except KeyboardInterrupt:
            logger.info("Programa interrompido pelo usuário")
        finally:
            client.disconnect()
    else:
        logger.error("Não foi possível conectar ao servidor")

if __name__ == "__main__":
    main()
